server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +415,301 @@ public int hashCode() {
+            }
+            if (canSkipCacheCleanup(stalenessThreshold)) {
+                return;
+            }
+            // Contains CleanupKey objects with open shard but invalidated readerCacheKeyId.
+            final Set<CleanupKey> cleanupKeysFromOutdatedReaders = new HashSet<>();
+            // Contains CleanupKey objects of a closed shard.
+            final Set<Object> cleanupKeysFromClosedShards = new HashSet<>();
+
+            for (Iterator<CleanupKey> iterator = keysToClean.iterator(); iterator.hasNext();) {
+                CleanupKey cleanupKey = iterator.next();
+                iterator.remove();
+                if (cleanupKey.readerCacheKeyId == null || !cleanupKey.entity.isOpen()) {
+                    // null indicates full cleanup, as does a closed shard
+                    cleanupKeysFromClosedShards.add(((IndexShard) cleanupKey.entity.getCacheIdentity()).shardId());
+                } else {
+                    cleanupKeysFromOutdatedReaders.add(cleanupKey);
+                }
+            }
+
+            if (cleanupKeysFromOutdatedReaders.isEmpty() && cleanupKeysFromClosedShards.isEmpty()) {
+                return;
+            }
+
             for (Iterator<Key> iterator = cache.keys().iterator(); iterator.hasNext();) {
                 Key key = iterator.next();
-                if (currentFullClean.contains(key.shardId)) {
+                if (shouldRemoveKey(key, cleanupKeysFromOutdatedReaders, cleanupKeysFromClosedShards)) {
                     iterator.remove();
-                } else {
-                    // If the flow comes here, then we should have a open shard available on node.
-                    if (currentKeysToClean.contains(
-                        new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId)
-                    )) {
-                        iterator.remove();
-                    }
                 }
             }
+            cache.refresh();
+        }
+
+        /**
+         * Determines whether a key should be removed from the cache.
+         *
+         * <p>This method checks if the key's shardId is present in the cleanupKeysFromClosedShards set,
+         * indicating that the shard has been closed and the key should be removed. If the shardId is not present,
+         * it checks if the key's readerCacheKeyId is present in the cleanupKeysFromOutdatedReaders set,
+         * indicating that the reader has been invalidated and the key should be removed.
+         *
+         * @param key The key to check for removal.
+         * @param cleanupKeysFromOutdatedReaders A set of CleanupKeys with open shard but invalidated readerCacheKeyId.
+         * @param cleanupKeysFromClosedShards A set of CleanupKeys of a closed shard.
+         * @return true if the key should be removed, false otherwise.
+         */
+        private synchronized boolean shouldRemoveKey(
```

Comment:
We should merge this logic in caller. It is just checking in 2 sets so not a complex logic to refactor in separate function.

Reply:
it just makes the code more readable instead of one giant method like we had before.

Reply:
discussed offline, that this is not necessary since the caller setsup the sets and the helper method is just checking for the presence.